### PR TITLE
feat: port `client-connect`, `client-dispatch`, `client-errors` test to `node:test`

### PR DESCRIPTION
**Pull request by sosukesuzuki**
## This relates to...

https://github.com/nodejs/undici/issues/2267

## Status

- [ ] I have read and agreed to the [Developer's Certificate of Origin][cert]
- [ ] Tested
- [ ] Benchmarked (**optional**)
- [ ] Documented
- [ ] Review ready
- [ ] In review
- [ ] Merge ready

[cert]: https://github.com/nodejs/undici/blob/main/CONTRIBUTING.md

### CODE REVIEW

File: test/node-test/client-connect.js [hunk 1]
```diff
@@ -228,81 +238,87 @@ test('basic connect error', (t) => {
       socket.end(data)
     })
   })
-  t.teardown(server.close.bind(server))
+  t.after(server.close.bind(server))
 
   server.listen(0, async () => {
     const client = new Client(`http://localhost:${server.address().port}`)
-    t.teardown(client.close.bind(client))
+    t.after(() => client.close.bind(client)())
```
Comment (metcoder95) [inline test/node-test/client-connect.js:245]: Same

File: test/node-test/client-dispatch.js [hunk 2]
```diff
@@ -110,7 +112,7 @@ test('basic dispatch get', (t) => {
 
   server.listen(0, () => {
     const client = new Client(`http://localhost:${server.address().port}`)
-    t.teardown(client.close.bind(client))
+    t.after(() => client.close.bind(client)())
```
Comment (metcoder95) [inline test/node-test/client-dispatch.js:115]: Same

File: test/node-test/client-connect.js [hunk 3]
```diff
@@ -24,97 +25,101 @@ test('basic connect', (t) => {
...
```
Comment (metcoder95) [inline test/node-test/client-connect.js:32]: It seems the function is getting called instead of passed

File: test/node-test/client-errors.js [hunk 4]
```diff
@@ -734,64 +753,66 @@ test('GET errors body', (t) => {
...
   const server = createServer((req, res) => {
     res.end('asd')
   })
-  t.teardown(server.close.bind(server))
+  t.after(server.close.bind(server))
 
   server.listen(0, () => {
     const client = new Client(`http://localhost:${server.address().port}`)
-    t.teardown(client.close.bind(client))
+    t.after(() => client.close.bind(client)())
```
Comment (metcoder95) [inline test/node-test/client-errors.js:784]: Same

File: test/node-test/client-errors.js [hunk 5]
```diff
@@ -994,26 +1025,28 @@ test('queued request should fail on client destroy', (t) => {
...
   })
-  t.teardown(server.close.bind(server))
+  t.after(server.close.bind(server))
 
   server.listen(0, () => {
     const client = new Client(`http://localhost:${server.address().port}`, {
       pipelining: 3
     })
-    t.teardown(client.close.bind(client))
+    t.after(() => client.close.bind(client)())
```
Comment (metcoder95) [inline test/node-test/client-errors.js:1049]: Same

File: test/node-test/client-connect.js (hunk 3 above)
Comment (sosukesuzuki) [reply test/node-test/client-connect.js]: I think the two codes below are equivalent:

```js
t.after(() => client.close.bind(client)())
```

```js
t.after(() => {
  return client.close.bind(client)()
})
```

However, to make it easier to understand, I rewritten it as `t.after(() => { client.close() })`

[e44332a](https://github.com/nodejs/undici/pull/2591/commits/e44332a81d42201c224edc83f0c54f9fb6b47c88)

File: test/node-test/async_hooks.js [hunk 6]
```diff
@@ -193,7 +193,7 @@ test('async hooks pipeline handler', async (t) => {
 
   server.listen(0, () => {
     const client = new Client(`http://localhost:${server.address().port}`)
-    t.after(() => client.close.bind(client)())
+    t.after(() => { client.close() })
```
Comment (mcollina) [inline test/node-test/async_hooks.js:196]:
```suggestion
    t.after(() => { return client.close() })
```

**Review by mcollina — CHANGES_REQUESTED** [review]
close should be updated everywhere

File: test/node-test/async_hooks.js (hunk 6 above)
Comment (mcollina) [inline test/node-test/async_hooks.js:196]:
```suggestion
    t.after(() => { return client.close() })
```

**Review by mcollina — CHANGES_REQUESTED** [review]

File: test/node-test/client-dispatch.js [hunk 7]
```diff
@@ -74,33 +76,33 @@ test('dispatch invalid opts', (t) => {
...
+    p.strictEqual('GET', req.method)
+    p.strictEqual(`localhost:${server.address().port}`, req.headers.host)
+    p.strictEqual(undefined, req.headers.foo)
+    p.strictEqual('bar', req.headers.bar)
+    p.strictEqual('', req.headers.baz)
+    p.strictEqual(undefined, req.headers['content-length'])
     res.end('hello')
   })
-  t.teardown(server.close.bind(server))
+  t.after(server.close.bind(server))
```
Comment (mcollina) [inline test/node-test/client-dispatch.js:105]: Unfortunately this does not work the same. server.close is a callback API, while t.after expects returning a Promise. Can you write a tiny adapter to match the two?

**Review by mcollina — CHANGES_REQUESTED** [review]

File: test/node-test/client-dispatch.js (hunk 7 above)
Comment (sosukesuzuki) [reply test/node-test/client-dispatch.js:105]: What do you think about the interface below?

```js
// definition
function closeServerAsPromise(server) {
  return () => util.promisify(server.close.bind(server))()
}

// usage
t.after(closeServerAsPromise(server))
```

Comment (mcollina) [reply test/node-test/client-dispatch.js:105]: That works

Comment (sosukesuzuki) [reply test/node-test/client-dispatch.js:105]: I've created new PR that adds `closeServerAsPromise` helper for existing tests. https://github.com/nodejs/undici/pull/2604

**Comment by mcollina** [comment]
@sosukesuzuki are you still interested in finishing this?

**Comment by sosukesuzuki** [comment]
Yes

**Review by mcollina — APPROVED** [review]
lgtm

File: test/client-connect.js [hunk 8]
```diff
@@ -6,277 +6,7 @@ const http = require('http')
...
-      signal: 'error',
-      opaque: 'asd'
-    }, (err, { opaque }) => {
-      t.equal(opaque, 'asd')
-      t.type(err, errors.InvalidArgumentError)
-    })
-  })
-})
-
+// TODO: move to test/node-test/client-connect.js
```
Comment (metcoder95) [inline test/client-connect.js:9]: This stills valid?

**Review by metcoder95 — APPROVED** [review]

**Pull request merged**